tatanka/client/client.go

```diff
@@ -0,0 +1,846 @@
+		return msgjson.NewError(mj.ErrAuth, "bad sig")
+	}
+
+	switch msg.Type {
+	case msgjson.Request:
+		handle, found := c.requestHandlers[msg.Route]
+		if !found {
+			// DRAFT NOTE: We should pontentially be more permissive of unknown
+			// routes in order to support minor network upgrades that add new
+			// routes.
+			c.log.Errorf("tatanka node %s sent a request to an unknown route %q", peerID, msg.Route)
+			return msgjson.NewError(mj.ErrBadRequest, "what is route %q?", msg.Route)
+		}
+		c.handleRequest(tt, msg, handle)
+	case msgjson.Notification:
+		handle, found := c.notificationHandlers[msg.Route]
+		if !found {
+			// DRAFT NOTE: We should pontentially be more permissive of unknown
+			// routes in order to support minor network upgrades that add new
+			// routes.
+			c.log.Errorf("tatanka node %s sent a notification to an unknown route %q", peerID, msg.Route)
+			return msgjson.NewError(mj.ErrBadRequest, "what is route %q?", msg.Route)
+		}
+		handle(tt, msg)
+	default:
+		c.log.Errorf("tatanka node %s send a message with an unhandleable type %d", msg.Type)
+		return msgjson.NewError(mj.ErrBadRequest, "message type %d doesn't work for me", msg.Type)
+	}
+
+	return nil
+}
+
+func (c *TankaClient) sendResult(tt *tatanka, msgID uint64, result interface{}) error {
+	resp, err := msgjson.NewResponse(msgID, result, nil)
+	if err != nil {
+		return err
+	}
+	if err := c.send(tt, resp); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *TankaClient) handleRequest(tt *tatanka, msg *msgjson.Message, handle func(*tatanka, *msgjson.Message) *msgjson.Error) {
+	if msgErr := handle(tt, msg); msgErr != nil {
+		respMsg := mj.MustResponse(msg.ID, nil, msgErr)
+		if err := c.send(tt, respMsg); err != nil {
+			c.log.Errorf("Send error: %v", err)
+		}
+	}
+}
+
+func (c *TankaClient) Auth(peerID tanka.PeerID) error {
+	c.tankaMtx.RLock()
+	tt, found := c.tankas[peerID]
+	c.tankaMtx.RUnlock()
+	if !found {
+		return fmt.Errorf("cannot auth with unknown server %s", peerID)
+	}
+	return c.auth(tt)
+}
+
+func (c *TankaClient) auth(tt *tatanka) error {
+	connectMsg := mj.MustRequest(mj.RouteConnect, &mj.Connect{ID: c.peerID})
+	var cfg *mj.TatankaConfig
+	if err := c.request(tt, connectMsg, &cfg); err != nil {
+		return err
+	}
+	tt.config.Store(cfg)
+	return nil
+}
+
+func (c *TankaClient) tankaNodes() []*tatanka {
+	c.tankaMtx.RLock()
+	defer c.tankaMtx.RUnlock()
+	tankas := make([]*tatanka, 0, len(c.tankas))
+	for _, tanka := range c.tankas {
+		tankas = append(tankas, tanka)
+	}
+	return tankas
+}
+
+func (c *TankaClient) PostBond(bond *tanka.Bond) error {
+	msg := mj.MustRequest(mj.RoutePostBond, []*tanka.Bond{bond})
+	for _, tt := range c.tankaNodes() {
+		var res bool
+		if err := c.request(tt, msg, &res); err != nil {
+			c.log.Errorf("error sending bond to tatanka node %s", tt.peerID)
+			continue
+		}
+		return nil
+	}
+	return errors.New("failed to report bond to any tatanka nodes")
```

Comment:
Good question. Seems right that we should report it to any that we're connected to. I guess I thought that the servers would share this info too, since it's reputation data.